### PR TITLE
Fix throwing exception due to the representation strings are the same for different AggregationFunctionState types

### LINKED ISSUE #55655: Unexpected exception Comparing aggregate functions with different types: AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64) and AggregateFunction(stochasticLinearRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64)
**Describe what's wrong**

https://fiddle.clickhouse.com/a7468d06-8a1b-485d-8f26-1d0ed75c6d27

The query (simplified from a production issue)
```
select *
from (
		select (
				select stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number)
				from numbers(10)
			) as col1,
			(
				select stochasticLinearRegressionState(0.1, 0., 5, 'SGD')(number, number)
				from numbers(10)
			) as col2
		from numbers(1)
	);       
```
throws an exception,
```
Code: 43. DB::Exception: Comparing aggregate functions with different types: AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64) and AggregateFunction(stochasticLinearRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64): While processing identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col1, identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLinearRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col2. (ILLEGAL_TYPE_OF_ARGUMENT) (version 23.2.4.1.2310B907)
, server ClickHouseNode
```

**Does it reproduce on recent release?**
Yes.

**How to reproduce**
https://fiddle.clickhouse.com/a7468d06-8a1b-485d-8f26-1d0ed75c6d27

* Which ClickHouse server version to use
Latest

**Expected behavior**
no exception

**Error message and/or stacktrace**
```
Code: 43. DB::Exception: Comparing aggregate functions with different types: AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64) and AggregateFunction(stochasticLinearRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64): While processing identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col1, identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLinearRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col2. (ILLEGAL_TYPE_OF_ARGUMENT) (version 23.2.4.1.2310B907)
, server ClickHouseNode
```

image

**Additional context**

### PR DESCRIPTION
**Pull request by Alex-Cheng**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix throwing exception due to the representation strings are the same for different AggregationFunctionState types Closes #55655

### Validation
Validated by running query
```
select *
from (
		select (
				select stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number)
				from numbers(10)
			) as col1,
			(
				select stochasticLinearRegressionState(0.1, 0., 5, 'SGD')(number, number)
				from numbers(10)
			) as col2
		from numbers(1)
	); 
```

### CODE REVIEW

**Comment by Alex-Cheng** [comment]
The label should be `pr-bugfix` but it seems that I cannot edit labels.

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 21fc0b5e99157da6c45c74b485e6b134b0e6ec31 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55711/21fc0b5e99157da6c45c74b485e6b134b0e6ec31/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6544057891">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55711/21fc0b5e99157da6c45c74b485e6b134b0e6ec31/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

**Review by nickitat — APPROVED** [review]

**Comment by Alex-Cheng** [comment]
image

By the fix of the PR, the query as following will no longer throw exception.
```
SELECT *
FROM
(
    SELECT
        bitmapHasAny(bitmapBuild([toUInt8(1)]), (
            SELECT groupBitmapState(toUInt8(1))
        )) has1,
        bitmapHasAny(bitmapBuild([toUInt64(1)]), (
            SELECT groupBitmapState(toUInt64(2))
        )) has2
) SETTINGS allow_experimental_analyzer = 0;
```

**Comment by Alex-Cheng** [comment]
@nickitat CI test failed but I cannot find anything related to the PR. It seems that some test is likely to fail randomly.

**Review by nickitat — APPROVED** [review]

File: src/Parsers/ASTLiteral.cpp [hunk 1]
```diff
@@ -94,7 +94,15 @@ void ASTLiteral::appendColumnNameImpl(WriteBuffer & ostr) const
         }
         else
         {
-            String column_name = applyVisitor(FieldVisitorToString(), value);
+            String column_name;
+            if (value.getType() == Field::Types::AggregateFunctionState)
+            {
+                column_name = value.template get<AggregateFunctionStateData>().name + applyVisitor(FieldVisitorToString(), value);
```
[the remark this answers is not in the document]
Comment (Alex-Cheng) [reply src/Parsers/ASTLiteral.cpp:100]: IMO, the column name is not table's column name, but ASTLiteral's column name which is generated on the fly. Could you please give me some details about the scenario where there is backward incompatible issue.

**Comment by Alex-Cheng** [comment]
image
Some tests are failed with no reason. How can I resolve it?

**Comment by Alex-Cheng** [comment]
> _This is an automated comment for commit [5a77353](https://github.com/ClickHouse/ClickHouse/commit/5a773536a67534acbe96b70bcb59e987edc031be) with description of existing statuses. It's updated for the latest CI running_
> 
> [❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55711/5a773536a67534acbe96b70bcb59e987edc031be/ci_running.html) to open a full report in a separate page
> 
> Successful checks
> Check name	Description	Status
> Integration tests	The integration tests report. In parenthesis the package type is given, and in square brackets are the optional part/total tests	[❌ failure](https://s3.amazonaws.com/clickhouse-test-reports/55711/5a773536a67534acbe96b70bcb59e987edc031be/integration_tests__asan__[1_6].html)
> Stress test	Runs stateless functional tests concurrently from several clients to detect concurrency-related errors	[❌ failure](https://s3.amazonaws.com/clickhouse-test-reports/55711/5a773536a67534acbe96b70bcb59e987edc031be/stress_test__asan_.html)
> Upgrade check	Runs stress tests on server version from last release and then tries to upgrade it to the version from the PR. It checks if the new server can successfully startup without any errors, crashes or sanitizer asserts	[❌ failure](https://s3.amazonaws.com/clickhouse-test-reports/55711/5a773536a67534acbe96b70bcb59e987edc031be/upgrade_check__asan_.html)

I am just wondering why it says `Upgrade check` is failed while the details page shows "all tests are ok".  "Integration tests" has the same issue.

image

**Comment by Alex-Cheng** [comment]
`test_backup_restore_on_cluster/test_disallow_concurrency.py::test_concurrent_backups_on_different_nodes` fails but I cannot find anything to do with my PR.

**Comment by Alex-Cheng** [comment]
`test_backup_restore_on_cluster/test_disallow_concurrency.py::test_concurrent_backups_on_different_nodes` is always fail but nothing to do with my PR, I feel it is a flacky.  @nickitat

File: src/Parsers/ASTLiteral.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tavplubix) [reply src/Parsers/ASTLiteral.cpp:100]: I was trying to find an example, but seems like I found another bug:
```
localhost.localdomain :) select * from (select (SELECT stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number)
            FROM numbers(10)) from numbers(1))

SELECT *
FROM
(
    SELECT (
            SELECT stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number)
            FROM numbers(10)
        )
    FROM numbers(1)
)

Query id: 4d8fd156-47fc-4bce-b0b2-a33861f92f3e


0 rows in set. Elapsed: 0.017 sec. 

Received exception:
Code: 47. DB::Exception: Missing columns: '_subquery5' while processing query: 'SELECT _subquery5 FROM (SELECT (SELECT stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number) FROM numbers(10)) FROM numbers(1))', required columns: '_subquery5' '_subquery5'. (UNKNOWN_IDENTIFIER)

```

But afaik in distributed queries we use column names that are generated on the fly for intermediate expression, and blocks with these names are transferred between nodes

[the remark this answers is not in the document]
Comment (nickitat) [reply src/Parsers/ASTLiteral.cpp:100]: @Alex-Cheng it is not hard to check. you could write an integration test that will spawn couple of nodes of different versions. e.g. https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/test_backward_compatibility/test_aggregate_function_state.py

[the remark this answers is not in the document]
Comment (Alex-Cheng) [reply src/Parsers/ASTLiteral.cpp:100]: > I was trying to find an example, but seems like I found another bug:
> 
> ```
> localhost.localdomain :) select * from (select (SELECT stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number)
>             FROM numbers(10)) from numbers(1))
> 
> SELECT *
> FROM
> (
>     SELECT (
>             SELECT stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number)
>             FROM numbers(10)
>         )
>     FROM numbers(1)
> )
> 
> Query id: 4d8fd156-47fc-4bce-b0b2-a33861f92f3e
> 
> 
> 0 rows in set. Elapsed: 0.017 sec. 
> 
> Received exception:
> Code: 47. DB::Exception: Missing columns: '_subquery5' while processing query: 'SELECT _subquery5 FROM (SELECT (SELECT stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number) FROM numbers(10)) FROM numbers(1))', required columns: '_subquery5' '_subquery5'. (UNKNOWN_IDENTIFIER)
> ```
> 
> But afaik in distributed queries we use column names that are generated on the fly for intermediate expression, and blocks with these names are transferred between nodes
Yes, this bug exists in latest version ,  https://fiddle.clickhouse.com/83b97a6b-4959-4fa9-a9bb-bbe7d00ed9e3.

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/pull/60136

**Pull request closed without merging**